File: app/src/main/java/com/softwareverde/tidyduck/most/MostInterface.java
package com.softwareverde.tidyduck.most;

import com.softwareverde.mostadapter.Modification;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class MostInterface {
    private Long _id;
    private String _mostId;
    private String _name;
    private String _description;
    private String _version;
    private Date _lastModifiedDate;
    private boolean _isDeleted;
    private Date _deletedDate;
    private boolean _isPermanentlyDeleted;
    private Date _permanentlyDeletedDate;
    private boolean _isApproved;
    private Long _approvalReviewId;
    private boolean _hasApprovedParent;
    private boolean _isReleased;
    private Long _baseVersionId;
    private Long _priorVersionId;
    private Long _creatorAccountId;

    private List<Modification> _modifications = new ArrayList<>();
    private List<MostFunction> _mostFunctions = new ArrayList<>();

    public Long getId() {
        return _id;
    }

    public void setId(Long id) {
        _id = id;
    }

    public String getMostId() {
        return _mostId;
    }

    public void setMostId(String mostId) {
        _mostId = mostId;
    }

    public String getName() {
        return _name;
    }

    public void setName(String name) {
        _name = name;
    }

    public String getDescription() {
        return _description;
    }

    public void setDescription(String description) {
        _description = description;
    }

    public String getVersion() {
        return _version;
    }

    public void setVersion(String version) {
        _version = version;
    }

    public Date getLastModifiedDate() {
        return _lastModifiedDate;
    }

    public void setLastModifiedDate(Date lastModifiedDate) {
        _lastModifiedDate = lastModifiedDate;
    }

    public boolean isDeleted() {
        return _isDeleted;
    }

    public void setIsDeleted(final boolean deleted) {
        _isDeleted = deleted;
    }

    public Date getDeletedDate() {
        return _deletedDate;
    }

    public void setDeletedDate(final Date deletedDate) {
        _deletedDate = deletedDate;
    }

    public boolean isPermanentlyDeleted() {
        return _isPermanentlyDeleted;
    }

    public void setIsPermanentlyDeleted(final boolean permanentlyDeleted) {
        _isPermanentlyDeleted = permanentlyDeleted;
    }

    public Date getPermanentlyDeletedDate() {
        return _permanentlyDeletedDate;
    }

    public void setPermanentlyDeletedDate(final Date permanentlyDeletedDate) {
        _permanentlyDeletedDate = permanentlyDeletedDate;
    }

    public boolean isApproved() {
        return _isApproved;
    }

    public void setIsApproved(final boolean approved) {
        _isApproved = approved;
    }

    public Long getApprovalReviewId() { return _approvalReviewId; }

    public void setApprovalReviewId(final Long approvalReviewId) { _approvalReviewId = approvalReviewId; }

    public boolean getHasApprovedParent() { return _hasApprovedParent; }

    public void setHasApprovedParent(final boolean hasApprovedParent) { _hasApprovedParent = hasApprovedParent; }

    public boolean isReleased() {
        return _isReleased;
    }

    public void setIsReleased(boolean released) {
        _isReleased = released;
    }

    public Long getBaseVersionId() {
        return _baseVersionId;
    }

    public void setBaseVersionId(Long baseVersionId) {
        _baseVersionId = baseVersionId;
    }

    public Long getPriorVersionId() {
        return _priorVersionId;
    }

    public void setPriorVersionId(Long priorVersionId) {
        _priorVersionId = priorVersionId;
    }

    public Long getCreatorAccountId() { return _creatorAccountId; }

    public void setCreatorAccountId(final Long creatorAccountId) { _creatorAccountId = creatorAccountId; }

    public List<Modification> getModifications() {
        return _modifications;
    }

    public void addModification(final Modification modification) {
        _modifications.add(modification);
    }

    public List<MostFunction> getMostFunctions() {
        return new ArrayList<>(_mostFunctions);
    }

    public void addMostFunction(MostFunction mostFunction) {
        _mostFunctions.add(mostFunction);
    }

    public void setMostFunctions(List<MostFunction> mostFunctions) {
        _mostFunctions = new ArrayList<>(mostFunctions);
    }
}

File: www/js/classes/most-interface.js
class MostInterface {
    static fromJson(json) {
        const mostInterface = new MostInterface();

        mostInterface.setId(json.id);
        mostInterface.setMostId(json.mostId);
        mostInterface.setName(json.name);
        mostInterface.setDescription(json.description);
        mostInterface.setLastModifiedDate(json.lastModifiedDate);
        mostInterface.setReleaseVersion(json.releaseVersion);
        mostInterface.setBaseVersionId(json.baseVersionId);
        mostInterface.setPriorVersionId(json.priorVersionId);
        mostInterface.setIsReleased(json.isReleased);
        mostInterface.setIsApproved(json.isApproved);
        mostInterface.setApprovalReviewId(json.approvalReviewId);
        mostInterface.setHasApprovedParent(json.hasApprovedParent);
        mostInterface.setCreatorAccountId(json.creatorAccountId);
        mostInterface.setIsDeleted(json.isDeleted);
        mostInterface.setDeletedDate(json.deletedDate);

        return mostInterface;
    }

    static toJson(mostInterface) {
        return {
            id:                 mostInterface.getId(),
            mostId:             mostInterface.getMostId(),
            name:               mostInterface.getName(),
            description:        mostInterface.getDescription(),
            lastModifiedDate:   mostInterface.getLastModifiedDate(),
            releaseVersion:     mostInterface.getReleaseVersion(),
            baseVersionId:      mostInterface.getBaseVersionId(),
            priorVersionId:     mostInterface.getPriorVersionId(),
            isReleased:         mostInterface.isReleased(),
            isApproved:         mostInterface.isApproved(),
            creatorAccountId:   mostInterface.getCreatorAccountId()
        };
    }

    constructor() {
        this._id                    = null;
        this._mostId                = null;
        this._name                  = "";
        this._description           = "";
        this._lastModifiedDate      = "";
        this._version               = "";
        this._versionsJson          = null;
        this._isReleased            = null;
        this._isApproved            = null;
        this._approvalReviewId      = null;
        this._hasApprovedParent     = false;
        this._priorVersionId        = null;
        this._baseVersionId         = null;
        this._creatorAccountId      = null;
        this._isDeleted             = false;
        this._deletedDate           = null;

        this._functions             = [];
    };

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setMostId(mostId) {
        this._mostId = mostId;
    }

    getMostId() {
        return this._mostId;
    }

    setName(name) {
        this._name = name;
    }

    getName() {
        return this._name;
    }

    setDescription(description) {
        this._description = description;
    }

    getDescription() {
        return this._description;
    }

    setLastModifiedDate(lastModifiedDate) {
        this._lastModifiedDate = lastModifiedDate;
    }

    getLastModifiedDate() {
        return this._lastModifiedDate;
    }

    setReleaseVersion(version) {
        this._version = version;
    }

    getReleaseVersion() {
        return this._version;
    }

    getFunctions() {
        return this._functions;
    }

    setVersionsJson(versionsJson) {
        this._versionsJson = versionsJson;
    }

    getVersionsJson() {
        return this._versionsJson;
    }

    setBaseVersionId(baseVersionId) {
        this._baseVersionId = baseVersionId;
    }

    getBaseVersionId() {
        return this._baseVersionId;
    }

    setPriorVersionId(priorVersionId) {
        this._priorVersionId = priorVersionId;
    }

    getPriorVersionId() {
        return this._priorVersionId;
    }

    setIsReleased(isReleased) {
        this._isReleased = isReleased;
    }

    isReleased() {
        return this._isReleased;
    }

    setIsApproved(isApproved) {
        this._isApproved = isApproved;
    }

    isApproved() {
        return this._isApproved;
    }

    setApprovalReviewId(approvalReviewId) {
        this._approvalReviewId = approvalReviewId;
    }

    getApprovalReviewId() {
        return this._approvalReviewId;
    }

    setHasApprovedParent(hasApprovedParent) {
        this._hasApprovedParent = hasApprovedParent;
    }

    hasApprovedParent() {
        return this._hasApprovedParent;
    }

    setCreatorAccountId(creatorAccountId) {
        this._creatorAccountId = creatorAccountId;
    }

    getCreatorAccountId() {
        return this._creatorAccountId;
    }

    setIsDeleted(isDeleted) {
        this._isDeleted = isDeleted;
    }

    isDeleted() {
        return this._isDeleted;
    }

    setDeletedDate(deletedDate) {
        this._deletedDate = deletedDate;
    }

    getDeletedDate() {
        return this._deletedDate;
    }

    getDisplayVersion() {
        if (this._isReleased) {
            return this._version;
        }
        return this._version + "-" + this._id;
    }
}
